### ISSUE TITLE
2022/01/21 - 2022/01/27

### ISSUE THREAD

**Issue by danbi5228** [njs03332/nlp_study#17]
* 1월 27일 pm 9:00
* 다음 진도
   * 유리님: 5.2
   * 선미님: 5.3
   * 단비님: 4.2 끝까지!

**Comment by danbi5228** [comment]
~~사실 앞서 정리와 논의를 많이 한 챕터라 그런지 4.2까지는 딱히 언급할 내용이 거의 없는 것 같아요? ㅎ~~

### Optimizer (추가 정리 필요)
- adam optimizer를 자연스럽게 쓰다가 이걸 왜쓰게되었더라... 하는 의문으로부터 시작됨

#### 3.4.5 옵티마이저 선택
```python

# 현재 거의 디폴트로 사용하는 옵티마이저
optimizer = optim.Adam(params=perceptron.parameter(), lr=lr)

# SGD (어려운 최적화 문제에 있어서 종종 나쁜모델을 만드는 경우가 있음)
optimizer = optim.SGD(perceptron.parameter(), lr=lr)
``` 

참고
* https://hiddenbeginner.github.io/deeplearning/2019/09/22/optimization_algorithms_in_deep_learning.html
* https://anweh.tistory.com/22

**Comment by givitallugot** [comment]
### 5.3 NewsClassifier

glove 임베딩 사용하여 뉴스 기사 제목을 바탕으로 클래스(스포츠, 과학/기술, 세계, 비즈니스) 분류

#### Vocabulary 클래스를 상속한 SequenceVocabulary
- SequenceVocabulary에서 특수 토큰 4개(UNK, MASK, BEGIN-OF-SEQUENCE, END-OF-SEQUENCE)를 추가
```python
class Vocabulary(object):
    def __init__(self, token_to_idx=None):
    ...

class SequenceVocabulary(Vocabulary):
    def __init__(self, token_to_idx=None, unk_token="<UNK>",
                         mask_token="<MASK>", begin_seq_token="<BEGIN>",
                         end_seq_token="<END>"):

        super(SequenceVocabulary, self).__init__(token_to_idx)
        # super(하위클래스의 이름, 하위클래스의 객체) 로 상위 클래스를 호출
```

#### 전이학습
- 사전 훈련된 임베딩을 로드
- 현재 데이터셋에 있는 단어에 해당하는 임베딩 일부만 추출
- Embedding 층의 가중치 행렬을 선택한 임베딩으로 지정
```python
class NewsClassifier(nn.Module):
    def __init__(pretrained_embeddings=None, ...):
        if pretrained_embeddings is None:
             ...
        else:
            pretrained_embeddings = torch.from_numpy(pretrained_embeddings).float()
            self.emb = nn.Embedding(embedding_dim=embedding_size,
                                    num_embeddings=num_embeddings,
                                    padding_idx=padding_idx,
                                    _weight=pretrained_embeddings) # _weight를 지정하면 pretrained 이용하는 것
    ...

classifier = NewsClassifier(...
                            pretrained_embeddings=embeddings) # 여기서 embeddings는 필요한 단어들만 선택된 glove 임베딩

```
- 만약 현재 데이터셋에 있는 단어 중 임베딩에 없는 경우 -> Xavier 균등 분포 등의 초기화 방법 사용
```python
  torch.nn.init.xavier_uniform_(embedding_i)
```
- 참고: https://blog.naver.com/PostView.nhn?blogId=hongjg3229&logNo=221564537122

**Comment by njs03332** [comment]
### 5.1.2 단어 임베딩 학습 방법
- Word2Vec의 CBoW vs. skip-gram 모델
  - <img width="602" alt="Screen Shot 2022-01-27 at 8 57 54 PM" src="https://user-images.githubusercontent.com/37107740/151354518-76a041b7-4a23-49cb-9606-602e9fc940e7.png">
  - CBoW (Continuous Bag of Word)
    - 주변 단어를 통해 주어진 단어를 예측
    - skip-gram 보다 속도가 빠름
    - 빈도가 높은 단어에 대해 조금 더 성능이 좋음
  - skip-gram
    - 중심 단어를 통해 주변 단어를 예측
    - 작은 데이터에 대해서도 CBoW 보다 일반적으로 성능이 좋아 더 많이 쓰임
    - 빈도가 낮은 단어/어절에 대해서도 학습 성능이 좋음
    - <img width="667" alt="Screen Shot 2022-01-27 at 8 59 06 PM" src="https://user-images.githubusercontent.com/37107740/151354724-cb4a7de4-7a4c-4c4b-b59f-956ebe31bcc2.png">
 
참조
https://towardsdatascience.com/nlp-101-word2vec-skip-gram-and-cbow-93512ee24314
https://simonezz.tistory.com/35
